File: backend/controller/ReservationController.java
package com.eventreservation.controller;

import com.eventreservation.dto.ReservationDto;
import com.eventreservation.service.ReservationService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/reservations")
@CrossOrigin(origins = "http://localhost:3000")
public class ReservationController {
    
    @Autowired
    private ReservationService reservationService;
    
    @PostMapping
    public ResponseEntity<?> createReservation(@Valid @RequestBody ReservationDto reservationDto, 
                                             Authentication authentication) {
        try {
            String userEmail = authentication.getName();
            ReservationDto createdReservation = reservationService.createReservation(reservationDto, userEmail);
            return ResponseEntity.status(HttpStatus.CREATED).body(createdReservation);
        } catch (RuntimeException e) {
            Map<String, String> error = new HashMap<>();
            error.put("error", e.getMessage());
            return ResponseEntity.badRequest().body(error);
        }
    }
    
    @GetMapping
    public ResponseEntity<List<ReservationDto>> getUserReservations(Authentication authentication) {
        String userEmail = authentication.getName();
        List<ReservationDto> reservations = reservationService.getUserReservations(userEmail);
        return ResponseEntity.ok(reservations);
    }
    
    @GetMapping("/{id}")
    public ResponseEntity<?> getReservationById(@PathVariable Long id, Authentication authentication) {
        String userEmail = authentication.getName();
        Optional<ReservationDto> reservation = reservationService.getReservationById(id, userEmail);
        
        if (reservation.isPresent()) {
            return ResponseEntity.ok(reservation.get());
        }
        
        Map<String, String> error = new HashMap<>();
        error.put("error", "Reservation not found or access denied");
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
    }
    
    @GetMapping("/code/{reservationCode}")
    public ResponseEntity<?> getReservationByCode(@PathVariable String reservationCode, 
                                                 Authentication authentication) {
        String userEmail = authentication.getName();
        Optional<ReservationDto> reservation = reservationService.getReservationByCode(reservationCode, userEmail);
        
        if (reservation.isPresent()) {
            return ResponseEntity.ok(reservation.get());
        }
        
        Map<String, String> error = new HashMap<>();
        error.put("error", "Reservation not found or access denied");
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
    }
    
    @PutMapping("/{id}/cancel")
    public ResponseEntity<?> cancelReservation(@PathVariable Long id, Authentication authentication) {
        try {
            String userEmail = authentication.getName();
            ReservationDto cancelledReservation = reservationService.cancelReservation(id, userEmail);
            return ResponseEntity.ok(cancelledReservation);
        } catch (RuntimeException e) {
            Map<String, String> error = new HashMap<>();
            error.put("error", e.getMessage());
            return ResponseEntity.badRequest().body(error);
        }
    }
    
    // Admin endpoints
    @GetMapping("/all")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<List<ReservationDto>> getAllReservations() {
        List<ReservationDto> reservations = reservationService.getAllReservations();
        return ResponseEntity.ok(reservations);
    }
    
    @GetMapping("/event/{eventId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<List<ReservationDto>> getEventReservations(@PathVariable Long eventId) {
        List<ReservationDto> reservations = reservationService.getEventReservations(eventId);
        return ResponseEntity.ok(reservations);
    }
}

File: backend/controller/reservationController.js
const express = require('express');
const { body, validationResult } = require('express-validator');
const reservationService = require('../service/reservationService');
const { protect, authorize } = require('../config/auth');

const router = express.Router();

// Validation rules
const reservationValidation = [
  body('event').isMongoId().withMessage('Valid event ID is required'),
  body('numberOfAttendees').isInt({ min: 1 }).withMessage('Number of attendees must be at least 1'),
  body('attendees').isArray({ min: 1 }).withMessage('At least one attendee is required'),
  body('attendees.*.name').trim().isLength({ min: 1 }).withMessage('Attendee name is required'),
  body('attendees.*.email').isEmail().withMessage('Valid attendee email is required')
];

// @desc    Create new reservation
// @route   POST /api/reservations
// @access  Private
router.post('/', protect, reservationValidation, async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        message: 'Validation failed',
        errors: errors.array()
      });
    }

    const reservation = await reservationService.createReservation(req.body, req.user._id);
    
    res.status(201).json({
      success: true,
      data: reservation,
      message: 'Reservation created successfully'
    });
  } catch (error) {
    res.status(400).json({
      success: false,
      message: error.message
    });
  }
});

// @desc    Get user's reservations
// @route   GET /api/reservations/my-reservations
// @access  Private
router.get('/my-reservations', protect, async (req, res) => {
  try {
    const result = await reservationService.getUserReservations(req.user._id, req.query);
    
    res.status(200).json({
      success: true,
      data: result.reservations,
      pagination: result.pagination
    });
  } catch (error) {
    res.status(400).json({
      success: false,
      message: error.message
    });
  }
});

// @desc    Get single reservation
// @route   GET /api/reservations/:id
// @access  Private
router.get('/:id', protect, async (req, res) => {
  try {
    const reservation = await reservationService.getReservationById(req.params.id, req.user._id);
    
    res.status(200).json({
      success: true,
      data: reservation
    });
  } catch (error) {
    res.status(404).json({
      success: false,
      message: error.message
    });
  }
});

// @desc    Update reservation
// @route   PUT /api/reservations/:id
// @access  Private
router.put('/:id', protect, async (req, res) => {
  try {
    const reservation = await reservationService.updateReservation(req.params.id, req.body, req.user._id);
    
    res.status(200).json({
      success: true,
      data: reservation,
      message: 'Reservation updated successfully'
    });
  } catch (error) {
    res.status(400).json({
      success: false,
      message: error.message
    });
  }
});

// @desc    Cancel reservation
// @route   DELETE /api/reservations/:id
// @access  Private
router.delete('/:id', protect, async (req, res) => {
  try {
    const { reason } = req.body;
    const result = await reservationService.cancelReservation(req.params.id, req.user._id, reason);
    
    res.status(200).json({
      success: true,
      message: result.message
    });
  } catch (error) {
    res.status(400).json({
      success: false,
      message: error.message
    });
  }
});

// @desc    Get reservations for an event (Organizer only)
// @route   GET /api/reservations/event/:eventId
// @access  Private (Organizer/Admin)
router.get('/event/:eventId', protect, authorize('organizer', 'admin'), async (req, res) => {
  try {
    const reservations = await reservationService.getEventReservations(req.params.eventId, req.user._id);
    
    res.status(200).json({
      success: true,
      data: reservations
    });
  } catch (error) {
    res.status(400).json({
      success: false,
      message: error.message
    });
  }
});

// @desc    Check in reservation (Organizer only)
// @route   PUT /api/reservations/:id/checkin
// @access  Private (Organizer/Admin)
router.put('/:id/checkin', protect, authorize('organizer', 'admin'), async (req, res) => {
  try {
    const result = await reservationService.checkInReservation(req.params.id, req.user._id);
    
    res.status(200).json({
      success: true,
      message: result.message
    });
  } catch (error) {
    res.status(400).json({
      success: false,
      message: error.message
    });
  }
});

// @desc    Get reservation by code
// @route   GET /api/reservations/code/:code
// @access  Private (Organizer/Admin)
router.get('/code/:code', protect, authorize('organizer', 'admin'), async (req, res) => {
  try {
    const reservation = await reservationService.getReservationByCode(req.params.code);
    
    res.status(200).json({
      success: true,
      data: reservation
    });
  } catch (error) {
    res.status(404).json({
      success: false,
      message: error.message
    });
  }
});

// @desc    Confirm payment
// @route   PUT /api/reservations/:id/confirm-payment
// @access  Private (Admin or payment webhook)
router.put('/:id/confirm-payment', protect, async (req, res) => {
  try {
    const result = await reservationService.confirmPayment(req.params.id, req.body);
    
    res.status(200).json({
      success: true,
      message: result.message
    });
  } catch (error) {
    res.status(400).json({
      success: false,
      message: error.message
    });
  }
});

// @desc    Get reservation statistics (Organizer only)
// @route   GET /api/reservations/stats/overview
// @access  Private (Organizer/Admin)
router.get('/stats/overview', protect, authorize('organizer', 'admin'), async (req, res) => {
  try {
    const stats = await reservationService.getReservationStats(req.user._id);
    
    res.status(200).json({
      success: true,
      data: stats
    });
  } catch (error) {
    res.status(400).json({
      success: false,
      message: error.message
    });
  }
});

module.exports = router;